### PR TITLE
Deprecate `nnet` and `signal` modules

### PR DESCRIPTION
**Pull request by rlouf**
In this PR we begin phasing the `nnet` and `signal` modules outside of Aesara, following discussion in https://github.com/aesara-devs/aesara/issues/674.

- [x] Disable auto-importing from `aesara.tensor`
- [x] Remove `aesara.tensor.nnet` and `aesara.tensor.signal` tests
- [x] Add deprecation warnings in `nnet` module 
- [x] Add deprecation warnings in `signal` module 
- [x] Move `Softmax` out of `aesara.tensor.nnet.basic` -> `aesara.tensor.math`
- [x] Move `LogSoftmax` out of `aesara.tensor.nnet.basic` -> `aesara.tensor.math`
- [x] Move `SoftmaxGrad` out of `aesara.tensor.nnet` -> `aesara.tensor.math`
- [x] Move tests for `Softmax`, `LogSoftmax`, `SoftmaxGrad`
- [x] Move rewrites related to `Softmax`, `LogSoftmax`, `SoftmaxGrad`
- [x] Rename `logsoftmax` -> `log_softmax`
- [x] Remove `nnet` and `signal` from the documentation, and any mention of anything NN-related

Anything else we want to salvage ?

### CODE REVIEW

**Comment by twiecki** [comment]
I would vote to also keep `conv2d` around, as convolutions are super common in all kinds of areas. I'd actually love to have a `conv1d` also, as there were too many cases where we had to roll our own.

**Comment by brandonwillard** [comment]
> Anything else we want to salvage ?

Looks good to me.

**Comment by ricardoV94** [comment]
Agree with @twiecki to keep the `conv2d` `Op` until we find some place to move it to. Everything else seems very specialized, easy to recreate.

**Comment by rlouf** [comment]
I have moved `Softmax`, `LogSoftmax` and `SoftmaxGrad` to `aesara.tensor.math`, but it may be better to create a new `aesara.tensor.special` module for these functions to have a familiar import path.

**Comment by rlouf** [comment]
Alright went through the checklist, will take another good look (especially at the documentation) tomorrow and move softmax-related `Op`s and functions to `aesara.tensor.special`. Then it should be good to merge once tests pass here.

Also, I indicated rel-2.9.0 as the deadline after which everything will be deleted. Does that sound right?

**Comment by brandonwillard** [comment]
> Also, I indicated rel-2.9.0 as the deadline after which everything will be deleted. Does that sound right?

Yeah, that sounds fine.

**Comment by twiecki** [comment]
So it seems like `conv2d` is being dropped?

**Comment by rlouf** [comment]
> I'd actually love to have a conv1d also, as there were too many cases where we had to roll our own.

We can open an issue to implement the equivalent of `np.convolve`. That's in scope.

As for conv_2d the question is more: why should we keep it _in Aesara_?

**Comment by rlouf** [comment]
Opened an issue for 1d convolution: https://github.com/aesara-devs/aesara/issues/1223

**Comment by brandonwillard** [comment]
> > conv_2d the question is more: why should we keep it in Aesara?
> 
> Don't have a good reason other than that it's useful to some PyMC models. The alternative is to add this to `pymc/aesaraf.py`, which is a fine solution too.

I believe the original idea was to create a separate project and move all this DL-specific code there.  We can always do that.  We can also keep `conv2d` for the time being, especially if removing it will break things in PyMC.

**Comment by rlouf** [comment]
> I believe the original idea was to create a separate project and move all this DL-specific code there.

Yeah that would be the ideal. You will still be able to import `conv_2d` after this PR is merged, and we can use the deprecation period to find a solution that works for everyone.

**Comment by rlouf** [comment]
I removed the tests since there's no point in having them in the codebase if we're going to skip them anyway (and they can be retrieved from the git history should we move them somewhere else), but I can bring them back if you think that's a better idea.

**Comment by brandonwillard** [comment]
> I removed the tests since there's no point in having them in the codebase if we're going to skip them anyway (and they can be retrieved from the git history should we move them somewhere else), but I can bring them back if you think that's a better idea.

Well, if we're going to keep the code around, we need to keep some tests.  We can consider reducing the amount of tests, because the tests in these sub-packages are among the most redundant, wasteful, and unnecessarily time-consuming.  That's actually one of the reasons why this refactoring is so immediately helpful.

**Comment by rlouf** [comment]
I reverted the commit that deleted the tests, here are all the `nnet` tests that take mode than 5s on my machine:

```
38.96s call     tests/tensor/nnet/test_abstract_conv.py::TestCorrConv3d::test_all
16.89s call     tests/tensor/nnet/test_abstract_conv.py::TestAbstractConvNoOptim::test_all
13.20s call     tests/tensor/nnet/test_batchnorm.py::test_batch_normalization
10.02s call     tests/tensor/nnet/test_batchnorm.py::test_bn_feature_maps
8.44s call     tests/tensor/nnet/test_batchnorm.py::test_batch_normalization_broadcastable
8.39s call     tests/tensor/nnet/test_conv.py::TestConv2D::test_invalid_input_shape
6.85s call     tests/tensor/nnet/test_abstract_conv.py::TestGroupedConv3dNoOptim::test_fwd
6.63s call     tests/tensor/nnet/test_abstract_conv.py::TestGroupedConv3dNoOptim::test_gradweights
6.45s call     tests/tensor/nnet/test_corr3d.py::TestGroupedConv3dNoOptim::test_fwd
6.13s call     tests/tensor/nnet/test_conv3d2d.py::test_conv3d[valid]
6.09s call     tests/tensor/nnet/test_corr3d.py::TestGroupedConv3dNoOptim::test_gradweights
5.94s call     tests/tensor/nnet/test_abstract_conv.py::TestCorrConv2d::test_gradinput_impossible_output_shapes
5.82s call     tests/tensor/nnet/test_abstract_conv.py::TestUnsharedConv::test_fwd
5.74s call     tests/tensor/nnet/test_corr.py::TestUnsharedConv::test_fwd
5.56s call     tests/tensor/nnet/test_conv.py::TestConv2D::test_missing_info
5.41s call     tests/tensor/nnet/test_batchnorm.py::test_batch_normalization_train_without_running_averages
5.33s call     tests/tensor/nnet/test_batchnorm.py::test_batch_normalization_train_broadcast
5.19s call     tests/tensor/nnet/test_corr3d.py::TestCorr3D::test_non_contiguous

```

**Comment by rlouf** [comment]
I am currently excluding all  the test related to `aesara.tensor.nnet` and `aesara.tensor.signal`. I can either include the tests related to `conv2d` individually, or use `pytests.mark.skip` on the others. I think refactoring the `nnet` and `signal` test suite is out of scope for this PR, and even for Aesara now that they're getting deprecated. 

What is your preferred solution?

**Review by brandonwillard — COMMENTED** [review]
Is there anything left on this one?  If not, we can probably merge this.

**Comment by rlouf** [comment]
Codecov warnings are about tests in `nnet` and `signal`, I think this can be reviewed one last time and merged.

File: aesara/scan/op.py
Comment (brandonwillard) [inline aesara/scan/op.py:3443]: This looks like it's in response to https://github.com/aesara-devs/aesara/issues/1221.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**
